zephyr/lib/alloc.c

```diff
@@ -68,12 +68,109 @@ extern char _end[], _heap_sentry[];
 
 static struct k_heap sof_heap;
 
+#if CONFIG_L3_HEAP
+static struct k_heap l3_heap;
+
+/**
+ * Returns the start of L3 memory heap.
+ * @return Pointer to the L3 memory location which can be used for L3 heap.
+ */
+static inline uintptr_t get_l3_heap_start(void)
+{
+	/*
+	 * TODO: parse the actual offset using:
+	 * - HfIMRIA1 register
+	 * - rom_ext_load_offset
+	 * - main_fw_load_offset
+	 * - main fw size in manifest
+	 */
+
+	return (uintptr_t)ROUND_UP(IMR_L3_HEAP_BASE, L3_MEM_PAGE_SIZE);
+}
+
+/**
+ * Returns the size of L3 memory heap.
+ * @return Size of the L3 memory region which can be used for L3 heap.
+ */
+static inline size_t get_l3_heap_size(void)
+{
+	 /*
+	  * Calculate the IMR heap size using:
+	  * - total IMR size
+	  * - IMR base address
+	  * - actual IMR heap start
+	  */
+	return ROUND_DOWN(IMR_L3_HEAP_SIZE, L3_MEM_PAGE_SIZE);
+}
+
+/**
+ * Checks whether pointer is from L3 heap memory range.
+ * @param ptr Pointer to memory being checked.
+ * @return True if pointer falls into L3 heap region, false otherwise.
+ */
+static bool is_l3_heap_pointer(void *ptr)
+{
+	uintptr_t l3_heap_start = get_l3_heap_start();
+	uintptr_t l3_heap_end = l3_heap_start + get_l3_heap_size();
+
+	if (POINTER_TO_UINT(ptr) >= l3_heap_start &&
+	    POINTER_TO_UINT(ptr) < l3_heap_end) {
+		return true;
+	}
+	return false;
```

Comment:
could be `return POINTER_TO_UINT(ptr) >= l3_heap_start && POINTER_TO_UINT(ptr) < l3_heap_end;`